README.md
# lua-zabbix-sender

A zabbix sender protocol implementation in Lua, for pushing monitoring data to Zabbix trapper items directly from your Lua code.

## Warning

I am not a programmer, so here’s a warning: **This code was written in an exploratory way**. If you encounter problems, see something wrong or something was implemented in a weird way, I would be happy if you tell me about it or create a pull request. Thank you. :)

## Example

```lua
local inspect = require('inspect')
local zbx_sender = require('zabbix-sender')
local sender = zbx_sender.new({
    server = 'localhost',
    port = 10051,
    monitored_host = 'node01'
  })

local resp = sender:add_item('trap1', 'test1')
  :add_item('trap2', 'test2')
  :add_item('trap1', 'test1', 'node02')
  :send()

print(inspect(resp))

{
  failed = 0,
  processed = 3,
  total = 3
}
```

```lua
local socket = require('socket')
local copas = require('copas')
local zbx_sender = require('zabbix-sender')

local sender = zbx_sender.new({
    server = 'localhost',
    port = 10051,
    monitored_host = 'node01',
    socket = function() return copas.wrap(socket.tcp()) end
  })

copas.addthread(function()
  sender:add_item('trap1', 'test1')
    :add_item('trap2', 'test2')

  copas.sleep(1)

  sender:add_item('trap1', 'test1', 'node02')
end)

copas.addthread(function()
  while sender:has_unsent_items() do
    sender:send()
    copas.sleep(2)
  end
end)

copas.loop()
```

## Documentation

* [Dependencies](#dependencies)
* [Functions](#functions)
   * [new](#newopts)
* [Methods](#methods)
   * [add_item](#add_itemkey-value-mhost)
   * [add_items](#add_itemsitems)
   * [clear](#clear)
   * [has_unsent_items](#has-unsent-items)
   * [send](#send)

## Dependencies

* lua >=5.1
* [luasocket](https://github.com/diegonehab/luasocket)
* [cjson](https://github.com/openresty/lua-cjson), [lunajson](https://github.com/grafi-tt/lunajson) or [dkjson](http://dkolf.de/src/dkjson-lua.fsl/home)
* [copas](https://github.com/keplerproject/copas) >=3.0.0 (**optional**)


## Functions

### `new([opts])`

Creates a new zabbix sender.

**Parameter:**

* *opts*: (**table**) options (**opt**)
* *opts.server*: (**string**) Zabbix server URL/IP (**default**=`localhost`)
* *opts.port*: (**number**) Zabbix server port (**default**=`10051`)
* *opts.monitored_host*: (**string**) The hostname the items belongs to. (**default**=`nil`)
* *opts.timestamps*: (**boolean**) Whether or not add timestamps to items - if `opts.nanoseconds` is set to `true` `opts.timestamps` is set to `true` too (**default**=`false`)
* *opts.nanoseconds*: (**boolean**) Whether or not add nanoseconds to items (**default**=`false`)
* *opts.timeout*: (**number**) connection timeout (**default**=`0.5`)
* *opts.socket*: (**function**) a function which returns a [LuaSocket TCP master object](https://w3.impa.br/~diego/software/luasocket/tcp.html). For example this can be used to wrap the socket with **copas** (See the example above). (**default**=`socket.tcp`)

**Returns:**

(**table**) zabbix sender

## Methods

### `add_item(key, value[, mhost])`

Adds an item to the request payload. The item(s) are stored until `send()` is invoked. Multiple calls to `add_item()` can be chained and completed by a final call to `send()`. See example above.

**Parameter:**

* *key*: (**string**) the items key
* *value*: (**string** | **number**) the items value
* *mhost*: (**string**) Is needed if [monitored_host](#newopts) was not set at creation or if the monitoring data is ment for another host.

**Returns:**

(**table**) self

**Raises:**

1. Error if `key` or `value` is missing
2. Error if `host` is missing and `monitored_host` was not set

### `add_items(items)`

Adds multiple items from a given table. The `items` table is an array of tables in the format of:

```lua
{
  { 'key1', 'val1', 'host1' },
  { 'key2', 'val2', 'host2' },
  ...
}
```

**Parameter:**

* *items*: (**table**) 

**Returns:**

(**table**) self

**Raises:**

1. Error if `items` is not given or not a `table`

### `clear()`

Removes all unsent items.

**Returns:**

(**self**)

### `has_unsent_items()`

Checks if there are currently unsent items.

**Returns:**

1. (**boolean**)

### `send()`

Sends all added items.

**Returns:**

1. (**table** | **boolean**) response from server | `false` in case of error
2. (**string**) error message in case of error


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [1.0.0-0] - 2021-11-24
### Added
- this changelog
- rockspec for dev
- three new options to `opts` table for `new()`
  - `opts.timestamps` to add timestamps to data
  - `opts.timeout` to set the connection timeout
  - `opts.socket` to wrap the socket for example
- added add_items() method

### Changed
- moved rockspec to rockspecs subfolder
- changed dependency for JSON module from `dkjson` to `lunajson`
- renamed two options in `opts` table for `new()`
  - `opts.host` changed to `opts.server`
  - `opts.with_ns` changed to `opts.nanoseconds`
- timestamps are no longer set by default, because they are not mandatory as I first thought (see `opts.timestamps`)
- the module now tries to load he following JSON modules: `cjson`, `lunajson`,`dkjson`
- has_unsent_items() no longer returns a second value
- clear() returns `self` now, instead of `nil`
- using exponentiation operator in `get_time`
- some internal renaming


### Fixed
- Used wrong variable name in line 123 (fixed by [evandrofisico](https://codeberg.org/evandrofisico) in [#4](https://codeberg.org/imo/lua-zabbix-sender/pulls/4))
- Client socket was not closed in case of error


## [0.1.0-0] - 2020-02-18
### Added
- First version


src/zabbix-sender.lua
local socket = require('socket')
local json do
    for _, mod in ipairs({'cjson', 'lunajson', 'dkjson'}) do
        local ok, j = pcall(require, mod)
        if ok then
            json = j
        end
    end

    assert(json, "Couldn't load any of the following JSON modules: cjson, lunajson, dkjson")
end

local zabbix_sender = {
    _VERSION      = '1.0.0-0',
    _DESCRIPTION  = 'A zabbix sender protocol implementation in Lua.',
    _URL          = 'https://git.kokolor.es/imo/lua-zabbix-sender',
    _LICENCE      = 'MIT'
}

-- protocol + flag
local ZHEAD_START = 'ZBXD\x01'

--- Privat functions

-- returns epoch and nanoseconds (I dunno how precise this is)
local function get_time()
    local time = socket.gettime()

    return math.floor(time), math.floor(( time % 1 ) * 10^9 )
end

-- set timestamp and nanoseconds if wanted
local function set_ts_and_ns(self, data)
    if self.timestamps then
        local ts, ns = get_time()
        data.clock = ts

        if self.nanoseconds then
            data.ns = ns
        end
    end
end

-- creates the payload as described in the docs
-- https://www.zabbix.com/documentation/5.0/manual/appendix/protocols/header_datalen
local function build_payload(self)
    local data = {
        request = 'sender data',
        data = self._items
    }
    set_ts_and_ns(self, data)
    self:clear()

    local jdata = json.encode(data)
    return string.format('%s%s%s', ZHEAD_START, string.pack('<L', jdata:len()), jdata)
end

-- parses the JSON response and returns only the data from then
-- info string
local function parse_response_data(resp)
    local resp_info = {}

    local ok, data = pcall(json.decode, resp)
    if not ok then
        return ok, data
    end

    data.info:gsub('(%w+):%s(%d+);', function(k, v)
        resp_info[k] = tonumber(v)
    end)

    return resp_info
end

local function receive_response(client)
    -- Note: maybe just use *a, because according to the docs the server
    -- closes the connection after it sent the status back
    local resp_head, err = client:receive(13)

    if not resp_head then
        client:close()
        return false, err
    elseif not resp_head:find('^' .. ZHEAD_START) or resp_head:len() ~= 13 then
        client:close()
        return false, 'Got invalid response from server'
    end

    local resp_data_len = string.unpack('<L', resp_head:sub(6))
    local resp_data = client:receive(resp_data_len)

    client:close()
    return parse_response_data(resp_data)
end


--- Main methods

local ZabbixSender = {}

function ZabbixSender:add_item(key, value, mhost)
    assert(key and value, 'Needs at least two arguments - key and value')
    mhost = mhost or self.monitored_host
    assert(mhost, 'No monitored host was given and no fallback was set')

    local item = {
        key = key,
        value = value,
        host = mhost
    }

    set_ts_and_ns(self, item)

    table.insert(self._items, item)

    return self
end

function ZabbixSender:add_items(items)
    assert(items and type(items) == 'table', 'Needs at least one argument - a table')

    for _, tbl in ipairs(items) do
        self:add_item(tbl[1], tbl[2], tbl[3])
    end

    return self
end

function ZabbixSender:clear()
    self._items = {}

    return self
end

function ZabbixSender:has_unsent_items()
    return #self._items ~= 0
end

function ZabbixSender:_connect()
    local client = self._socket()
    client:settimeout(self.timeout)

    local ok, err = client:connect(self.server, self.port)
    if not ok then
        client:close()
        return false, err
    end

    return client
end

function ZabbixSender:send()
    local items = self._items
    local data = build_payload(self)

    local client, err = self:_connect()
    if not client then
        self._items = items
        return client, err
    end

    local ok, err = client:send(data) -- luacheck: ignore 411/ok err
    if not ok then
        client:close()
        self._items = items
        return false, err
    end

    local resp, err = receive_response(client) -- luacheck: ignore 411/err
    if not resp then
        return resp, err
    end

    return resp
end


--- Public functions

function zabbix_sender.new(opts)
    opts = opts or {}
    local self = {
        server = opts.server or 'localhost',
        port = tonumber(opts.port) or 10051,
        monitored_host = opts.monitored_host,
        timestamps = opts.timestamps or opts.nanoseconds or false,
        nanoseconds = opts.nanoseconds or false,
        timeout = opts.timeout or 0.5,
        _socket = opts.socket or socket.tcp,
        _items = {}
    }

    return setmetatable(self, { __index = ZabbixSender })
end

return zabbix_sender
